torch/csrc/autograd/function.cpp

```diff
@@ -36,95 +36,53 @@ AnomalyMetadata* Function::metadata() noexcept {
- *    the memory of these stack frames will probably grow as time passes.
- * Testing on a few machines machines, the magic numbers were:
- * - Mac OSX (Macbook Pro 15) : ~60000
- * - A beefy Ubuntu 16.04 box : ~15000
- * - Windows AWS instance (g3.4xlarge): variable. My two attempts at different
- *   times have gotten the following numbers: ~8300, 3669
- */
-#ifdef _WIN32
-size_t deleteFunctionMaxRecursionDepth = 3000;
-#else
-size_t deleteFunctionMaxRecursionDepth = 10000;
-#endif
-
-struct RecursionDepthCounter {
- public:
-  explicit RecursionDepthCounter() {
-    ++deleteFunctionRecursionDepth;
-  }
-  ~RecursionDepthCounter() {
-    --deleteFunctionRecursionDepth;
-  }
-
-  size_t value() {
-    return deleteFunctionRecursionDepth;
+static void gatherFunctions(Function* func,
+                            std::vector<std::shared_ptr<Function>>& stack) {
+  for (auto& edge : func->next_edges()) {
+    if (edge.function.use_count() == 1) {
+      stack.emplace_back(std::move(edge.function));
+    }
   }
-};
+}
 
 /*
- * Note that the custom deleter deletes in BFS style. Without using
- * the custom deleter, the computation graph is deleted in a DFS style.
- * The BFS deletion is valid (and safe) because if a shared_ptr<Function>
- * 's reference count hits 0, nothing else will access it.
- */
+  * Fix for #5534: prevent stack overflow on deletion of deep computation graph
+  * 
+  * Sometimes one can end up with a very big computation graph of Functions
+  * and Edges. Each std::shared_ptr<Function> contains a list of Edge, and
+  * each Edge contains a std::shared_ptr<Function>. Deleting a
+  * std::shared_ptr<Function> can trigger the recursive deletion of other
+  * std::shared_ptr<Function>'s: this can stack overflow if the graph
+  * is deep enough. Here is an example of such a graph:
+  *
+  * shared_ptr<Function> -> Edge -> shared_ptr<Function> -> Edge -> ... -> shared_ptr<Function>
+  *
+  * The solution here is to detect when we are decrementing away the last
+  * reference to a Function, and when doing so to buffer up the Function's
+  * that will be recursively decremented.  We can then decrement (and free)
+  * the original Function without causing a recursive cascade, before
+  * draining the buffer applying the same behavior.  This is, in effect,
+  * converting recursion to a loop, using a heap buffer in place of the
+  * recursive call stack.
+  */
 void deleteFunction(Function* function) {
-  RecursionDepthCounter recursion_depth;
+  // To avoid stack overflow on large computational graphs,
+  // we need to track reference decrementing and freeing
+  // on the heap.
+  std::vector<std::shared_ptr<Function>> stack;
+  gatherFunctions(function, stack);
+  delete function;
 
-  if (recursion_depth.value() > deleteFunctionMaxRecursionDepth) {
-    deleteFunctionQueue.push_back(function);
-    return;
-  }
+  while (!stack.empty()) {
+    auto& curr_func = stack.back();
 
-  delete function;
+    if (curr_func.use_count() == 1) {
```

Comment:
I think `curr_func` is guaranteed to have `use_count = 1` here, right?

Reply:
Good point.  That wasn't true in an earlier version of this change, but is true now.

Reply:
https://github.com/pytorch/pytorch/pull/11601